python/ray/dag/compiled_dag_node.py

```diff
@@ -1075,6 +1192,202 @@ def _get_or_compile(
+                idx = operation_nodes[0].idx
+                read_node, compute_node, write_node = (
+                    operation_nodes[0],
+                    operation_nodes[1],
+                    operation_nodes[2],
+                )
+                # Add edges from READ to COMPUTE, and from COMPUTE to WRITE, which
+                # belong to the same task.
+                read_node.add_edge(compute_node)
+                compute_node.add_edge(write_node)
+                # Add an edge from COMPUTE with `bind_index` i to COMPUTE with
+                # `bind_index` i+1 if they belong to the same actor.
+                if prev_compute_node is not None:
+                    prev_compute_node.add_edge(compute_node)
+                prev_compute_node = compute_node
+                assert idx not in graph
+                graph[idx] = {
+                    DAGNodeOperationType.READ: read_node,
+                    DAGNodeOperationType.COMPUTE: compute_node,
+                    DAGNodeOperationType.WRITE: write_node,
+                }
+
+        from ray.dag import (
+            ClassMethodNode,
+            MultiOutputNode,
+        )
+
+        # Add an edge from WRITE of the writer task to READ of the reader task.
+        for idx, task in self.idx_to_task.items():
+            if not isinstance(task.dag_node, ClassMethodNode):
+                continue
+            for downstream_idx in task.downstream_node_idxs:
+                downstream_dag_node = self.idx_to_task[downstream_idx].dag_node
+                if isinstance(downstream_dag_node, MultiOutputNode):
+                    continue
+                graph[idx][DAGNodeOperationType.WRITE].add_edge(
+                    graph[downstream_idx][DAGNodeOperationType.READ]
+                )
+        return graph
+
+    def _build_execution_schedule(self):
+        """
+        Generate an execution schedule for each actor. The schedule is a list of
+        DAGNodeOperation.
+
+        Step 1: Generate a DAG node operation graph. Refer to the functions
+        `_generate_dag_operation_graph_node` and `_build_dag_node_operation_graph`
+        for more details.
+
+        Step 2: Topological sort
+
+        It is possible to have multiple DAGOperationGraphNodes with zero in-degree.
+        Refer to the function `_select_next_nodes` for the logic of selecting nodes.
+
+        Then, put the selected nodes into the corresponding actors' schedules.
+
+        The schedule should be intuitive to users, meaning that the execution should
+        perform operations in ascending order of `bind_index` as much as possible.
+
+        [Example]:
+
+        See `test_execution_schedule` for more examples.
+        """
+        # Step 1: Build a graph of DAGOperationGraphNode
+        actor_to_operation_nodes = self._generate_dag_operation_graph_node()
+        graph = self._build_dag_node_operation_graph(actor_to_operation_nodes)
+
+        actor_to_candidates: Dict[
```

Comment:
We can keep this current method here but let's separate out all of the logic from here down into a separate function in `dag_node_operation.py`. That will make the logic unit-testable without needing to reference the CompiledDAGNode class.

Reply:
Sounds good!

Reply:
Added:
* https://github.com/ray-project/ray/pull/46911/commits/44abfbbe7d1b82710fadedf04df97d0e82954688
* https://github.com/ray-project/ray/pull/46911/commits/bca959b33fe0109f6c7578073276fb1f08a6e1a3
* https://github.com/ray-project/ray/pull/46911/commits/c98362becce7f73499bc140b547f80731760d9cd